### ISSUE TITLE
Set object storage quota per namespace/project via a `ResourceQuota` resource

### ISSUE THREAD

**Issue by tumido** [rook/rook#7906]
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Cluster admins should be allowed to limit maximal total bucket sizes per project - no matter how many buckets are created in the project or what size limit is specified in the bucket claim.

**What is use case behind this feature:**
Currently any project can starve others out of the storage capacity, because there's no way (as far as I know, apologies if I missed something) to define the total limit on max size of all buckets in a namespace.

The only way to cap a bucket size is via `ObjectBucketClaim` which is in full control of the project admins. This doesn't account for mutliple buckets in a namespace and their total size. Additionally it is necessary to be able to enforce quotas in a way that project/namespace admins can't change them (hence the focus on `ResourceQuota` resource). 

**Environment**:
Openshift 4.7.7
Openshfit Container Storage 4.6
Rook v1.6.0

**Comment by travisn** [comment]
@thotz Is it possible to add the bucket quotas to the object storage class? From this [related discussion](https://github.com/rook/rook/pull/5674/files#r451930301) I thought we either had it or would be adding it.

**Comment by thotz** [comment]
@travisn : Idea behind `bucketquotas` to `storage class` is to limit the usage of bucket specified in `storageclass`(only valid for existing buckets). So that in case of multiple references to this bucket in different `OBC's/storage class` can be limited specifically. The current quota support for OBC is based on user scope.
Saying that @tumido request is more or like limiting usage for object storage(no of buckets and size I guess) but at least from the [docs](https://docs.ceph.com/en/latest/radosgw/admin/#quota-management), RGW quota scope is based only on `user` and `bucket`, you can default quota for all the users and buckets in the object store. But still, the no of users/buckets cannot be limited.
Will this help your use case?
WRT `Resourcequota` , the current `storagequota` implementation supports for `pvs` and storage class to related to `pv` which k8s itself have the knowledge and it can take care of itself. I guess the `ceph-csi driver` or `Rook` does not have a role in that management. We can get similar feature for `BucketRequest` or `BucketAccessRequest` for the COSI project(it will take time)

Another way is to work with `ResourceQuota` for this `storageclass` is `Rook` or `libbucket provisioner` handle that management. Say for example if we define 

```yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: storage-consumption
spec:
  hard:
    rook-ceph-delete-bucket.storageclass.storage.k8s.io/requests.storage: "10Gi" 
```
If the above `ResourceQuota` is defined, then while creating the first OBC out of this `storage class` it assigns quota with 10 GB unless the user specifies `MaxSize` in the OBC CRD.

Please note we cannot work with the global `requests.storage` since it is mainly for pvc's and no OBC cannot be limited since `ResourceQuota` does not have support for that object. Another riskier approach to limit the `secret` or `configmap` objects which can affect the creation of OBC since each OBC creates a `secret` or `configmap`.

We can also do this without using `ResourceQuota` in `Parameters` `StorageClass.yaml`. Again `RGW` cannot make any restrictions only Rook need to manage this case(or lib bucket provisioned library)

**Comment by travisn** [comment]
@thotz I'm not clear on the expected behavior. Here is the scenario:
- The admin wants to put a quota on buckets (the number of objects in the bucket and the size of the bucket)
- The admin would set the quota on the object storage class
- When the bucket is provisioned, the quota on the storage class would be applied. If a quota is specified on the OBC, it would be ignored since the admin has overridden the quota.

Basically, the admin should have control of the bucket quotas. Is this possible today? Or are you saying that quotas are only a user setting on the OBC? If it's only a user setting, I'm still not understanding why we have quotas since the user has no reason even to add quotas on their own buckets, as discussed [separately](https://github.com/rook/rook/pull/5674/files#r451930301).

**Comment by tumido** [comment]
Yeah, my usecase was mainly around "How can a cluster admin limit how much of the object storage can be consumed by a certain namespace?"

This is a big problem in multitenant clusters for us. As @travisn mentioned, users usually don't have any motivation to restrict themselves in the amount of consumed resources and the only way for us to maintain clusters in shape (not to starve other users from resources) is to set some limits per namespace. And we would like to do have a similar mechanism for object storage as we have for CPUs, memory and PVCs.

Also, `ResourceQuota` already supports limiting bucket counts via `count/objectbucketclaims` or `count/objectbuckets` (since this is handled by kube API I suppose). I think it would be nice if the same `ResourceQuota` can be extended to support the size/object count limits.

If I'm not mistaken, I think the per-storage-class-quota described by @thotz above does not apply, since that's also limited to PVCs/PVs.

About the other approach outlined above - setting the quota per Ceph user, can you please show me any example how to enforce that for users? In our case, users interface with a centralized rook/noobaa deployment and their only contact point for Rook is via `ObjectBucketClaim` resource. They don't need to do anything more than submit that single resource to get a bucket...

**Comment by thotz** [comment]
> Yeah, my usecase was mainly around "How can a cluster admin limit how much of the object storage can be consumed by a certain namespace?"
> 
> This is a big problem in multitenant clusters for us. As @travisn mentioned, users usually don't have any motivation to restrict themselves in the amount of consumed resources and the only way for us to maintain clusters in shape (not to starve other users from resources) is to set some limits per namespace. And we would like to do have a similar mechanism for object storage as we have for CPUs, memory and PVCs.
> 
> Also, `ResourceQuota` already supports limiting bucket counts via `count/objectbucketclaims` or `count/objectbuckets` (since this is handled by kube API I suppose). I think it would be nice if the same `ResourceQuota` can be extended to support the size/object count limits.

Good to know.

> 
> If I'm not mistaken, I think the per-storage-class-quota described by @thotz above does not apply, since that's also limited to PVCs/PVs.
> 
> About the other approach outlined above - setting the quota per Ceph user, can you please show me any example how to enforce that for users? In our case, users interface with a centralized rook/noobaa deployment and their only contact point for Rook is via `ObjectBucketClaim` resource. They don't need to do anything more than submit that single resource to get a bucket...

So I guess u have `ceph user` created using `cephobjectstoreuser` CRD in Rook and u need to limit quota for that user for the bucket, size etc? If yes u can raise RFE for that which can be done easily I guess.

**Comment by thotz** [comment]
> @thotz I'm not clear on the expected behavior. Here is the scenario:
> 
>     * The admin wants to put a quota on buckets (the number of objects in the bucket and the size of the bucket)
> 
>     * The admin would set the quota on the object storage class
> 
>     * When the bucket is provisioned, the quota on the storage class would be applied. If a quota is specified on the OBC, it would be ignored since the admin has overridden the quota.

This is a good approach much simpler than what in my mind(as parameter inherited to OBC and make much more sense) 

> 
> 
> Basically, the admin should have control of the bucket quotas. Is this possible today? Or are you saying that quotas are only a user setting on the OBC? If it's only a user setting, I'm still not understanding why we have quotas since the user has no reason even to add quotas on their own buckets, as discussed [separately](https://github.com/rook/rook/pull/5674/files#r451930301).

Sorry for the confusion, the admin has control over setting the quota on OBC, what I meant was internally RGW enforces user quota when admin sets quota on an OBC(on `user` created as part of OBC request).

What I was trying to communicate was, we cannot set quota on the object store usage(like no of buckets, size of the object store  consumption)

**Comment by thotz** [comment]
For getting a closure here, 
@travisn : I will work on quota proposed by you in storageclass of OBC
@tumido : I will work on the quota feature for `CephObjectstoreUser`

Please share if there is any concerns

**Comment by BlaineEXE** [comment]
In progress via https://github.com/rook/rook/pull/8426
